SYS_ATL/LoopIR_scheduling.py

```diff
@@ -1679,6 +1679,38 @@ def map_s(self, s):
             return ([single_stmt],[],[])
 
 
+class _DoRemoveLoop(LoopIR_Rewrite):
+    def __init__(self, proc, stmt):
+        assert isinstance(stmt, LoopIR.stmt)
+        self.stmt = stmt
+        super().__init__(proc)
+
+        self.proc = InferEffects(self.proc).result()
+
+    def map_s(self, s):
+        if s is self.stmt:
+            # Check if we can remove the loop
+            # Conditions are:
+            # 1. Body does not depend on the loop iteration variable
+            # 2. Body is idemopotent
+            # 3. The loop runs at least once
+            # we should run something similar to old effectcheck's
+            # is_pos_int to check 3, but that's a TODO.
+            # for now, we'll check 1 and 2.
```

Comment:
Not including check (3) makes this unsafe, which is not great. You should _make_ it safe by wrapping the loop body in an `if` that checks the loop extent is positive at runtime. It's possible the simplifier could (in some cases) resolve the condition to `True` and eliminate the branch. It's better to have a safe scheduling language that could be improved than an unsafe scheduling language that goes faster when it happens to work, sometimes.

Finally, TODO comments should be formatted as `# TODO: ...` to make them easy to search for. After implementing the runtime check, I suggest:

```
# TODO: (3) could be checked statically using something similar to the legacy is_pos_int.
```

Reply:
Well, `remove_loop` is already happening as part of `fission_after` without (3). We have `is_pos_int` in the old effect check, but that interface is not exposed to LoopIR_scheduling.py, and I think @gilbo is working on making a similar interface with new effect check. I will need to implement whole SMT solver logic here if I want to check (3), which will be a huge duplication.

Reply:
I'm saying to check (3) at _runtime_, which only involves the following rewrite:

```
for i in seq(0, N):  =>  if N > 0:
  idempotent body          idempotent body
```

Reply:
Ah I see! That makes sense, thanks!